app/src/main/java/net/kwatts/powtools/util/BluetoothUtilImpl.java

```diff
@@ -293,34 +299,46 @@ public void onScanResult(int callbackType, ScanResult result) {
         @Override
         public void onBatchScanResults(List<ScanResult> results) {
             for (ScanResult sr : results) {
-                Log.i(TAG,"ScanCallback.onBatchScanResults.each:" + sr.toString());
+                Timber.i("ScanCallback.onBatchScanResults.each:" + sr.toString());
             }
         }
 
         @Override
         public void onScanFailed(int errorCode) {
-            Log.e(TAG, "ScanCallback.onScanFailed:" + errorCode);
+            Timber.e( "ScanCallback.onScanFailed:" + errorCode);
         }
     };
 
 
     public void connectToDevice(BluetoothDevice device) {
-        Log.e(TAG, "connectToDevice:" + device.getName());
+        Timber.e( "connectToDevice:" + device.getName());
         device.connectGatt(mainActivity, false, mGattCallback);
     }
 
 
     private void onOWStateChangedToDisconnected(BluetoothGatt gatt) {
-        updateLog("We got disconnected from our Device: " + gatt.getDevice().getAddress());
+        Timber.i("We got disconnected from our Device: " + gatt.getDevice().getAddress());
         mainActivity.deviceConnectedTimer(false);
         mOWDevice.isConnected.set(false);
         App.INSTANCE.releaseWakeLock();
         mScanResults.clear();
 
         if (App.INSTANCE.getSharedPreferences().shouldAutoReconnect()) {
-            updateLog("Attempting to Reconnect to " + mOWDevice.deviceMacAddress.get());
-            BluetoothDevice device = mBluetoothAdapter.getRemoteDevice(mOWDevice.deviceMacAddress.get());
-            connectToDevice(device);
+            mRetryCount++;
+            try {
+                Timber.i("Sleeping for 5 seconds, mRetryCount=" + mRetryCount);
+                TimeUnit.SECONDS.sleep(5);
```

Comment:
Which thread is this on? We should probably start a new thread (or use a thread pool) to wait the 5 seconds.

Reply:
Good spot and the last of the things that need to be reviewed. I think it's on the BLE thread, and this does seem to work, although will continue to try and connect even when the app is in the background. The best approach may be what the official app does - continue to try and connect unless the app is actively closed by the user.